packages/typespec-vscode/src/emit/emit.ts

```diff
@@ -0,0 +1,395 @@
+      showOutput: true,
+      showPopup: true,
+      progress: overallProgress,
+    });
+    try {
+      const npmInstallResult = await npmUtil.npmInstallPackages(packagesToInstall, undefined, {
+        onStdioOut: toOutput,
+        onStdioError: toError,
+      });
+      if (npmInstallResult.exitCode !== 0) {
+        logger.error(`Error occurred when installing packages.`, [`${npmInstallResult.stderr}`], {
+          showOutput: true,
+          showPopup: true,
+          progress: overallProgress,
+        });
+        return;
+      }
+      logger.info("completed install...");
+    } catch (err) {
+      logger.error(`Exception occurred when installing packages.`, [err], {
+        showOutput: true,
+        showPopup: true,
+        progress: overallProgress,
+      });
+      return;
+    }
+  }
+
+  /* emit */
+  logger.info("Generating ...", [], {
+    showOutput: false,
+    showPopup: false,
+    progress: overallProgress,
+  });
+
+  const cli = await resolveTypeSpecCli(baseDir);
+  if (!cli) {
+    logger.error(
+      "Cannot find TypeSpec CLI. Please install @typespec/compiler. Generating Cancelled.",
+      [],
+      {
+        showOutput: true,
+        showPopup: true,
+        progress: overallProgress,
+      },
+    );
+    return;
+  }
+  const outputDir = path.resolve(baseDir, selectedEmitter.emitterKind, selectedEmitter.language);
+
+  const options: Record<string, string> = {};
+  options["emitter-output-dir"] = outputDir;
+  logger.info(
+    `Start to generate ${selectedEmitter.language} ${selectedEmitter.emitterKind} code under ${outputDir}...`,
+    [],
+    {
+      showOutput: true,
+      showPopup: false,
+      progress: overallProgress,
+    },
+  );
+  try {
+    const compileResult = await compile(cli, mainTspFile, selectedEmitter.package, options);
+    if (compileResult.exitCode !== 0) {
+      logger.error(
+        `Failed to generate ${selectedEmitter.emitterKind} code for ${selectedEmitter.language}.`,
+        [],
+        {
+          showOutput: true,
+          showPopup: true,
+          progress: overallProgress,
+        },
+      );
+    } else {
+      logger.info(
+        `Generating ${selectedEmitter.emitterKind} code for ${selectedEmitter.language}...Succeeded`,
+        [],
+        {
+          showOutput: true,
+          showPopup: true,
+          progress: overallProgress,
+        },
+      );
+    }
+  } catch (err) {
+    logger.error(
+      `Exception occurred when generating ${selectedEmitter.emitterKind} code for ${selectedEmitter.language}.`,
+      [err],
+      {
+        showOutput: true,
+        showPopup: true,
+        progress: overallProgress,
+      },
+    );
+  }
+
+  /*TODO: build sdk. */
+}
+
+export async function emitCode(
+  context: vscode.ExtensionContext,
+  uri: vscode.Uri,
+  overallProgress: vscode.Progress<{ message?: string; increment?: number }>,
+) {
+  let tspProjectFile: string = "";
```

Comment:
shall we check no compile error first? Emit would fail anyway when we can't compile successfully...

Reply:
We don't need to check compile here. Emit code does not pre-dependent with compiler directly. When emit code, we will install emitter package if not installed, in that way, the compiler will be installed together. So here, we don't need compiler be installed.

Reply:
I didn't mean the compiler not installed, i mean if there is compiler error reported already, we dont need to wait for the last step to tell user the compile failed. we can report it directly at very beginning. you should be able to get compiler error from vscode directly

Reply:
In current compiler scenario, we don't need to separate parser and emit into two different steps. tsp compile always parser + emitter code together if emitter is specified, and most case there is a emit define in tspconfig.yaml. and emit stopped if compile has failure.